Mathlib/Algebra/MonoidAlgebra/NoZeroDivisors.lean

```diff
@@ -161,4 +190,95 @@ theorem NoZeroDivisors.of_right_ordered [NoZeroDivisors R] [AddLeftCancelSemigro
 
 end LeftOrRightOrderability
 
+set_option autoImplicit false
+
+section Covariant_lt
+variable [Add A]
+
+section PartialOrder
+variable [PartialOrder A] [CovariantClass A A (· + ·) (· < ·)] {a t b : A}
+  {f g : AddMonoidAlgebra R A}
+
+/--  The "top" element of `AddMonoidAlgebra.single a r * f` is the product of `r` and
+the "top" element of `f`.  Here, "top" is simply an upper bound for the elements
+of the support of `f` (e.g. the product is `0` if "top" is not a maximum).
+The corresponding statement for a general product `f * g` is `AddMonoidAlgebra.mul_apply_of_le`.
+It is proved with a further `CovariantClass` assumption. -/
+theorem single_mul_apply_of_le (r : R) (ft : ∀ a ∈ f.support, a ≤ t) :
+  ((AddMonoidAlgebra.single a r) * f) (a + t) = r * f t := by
+  classical
+  simp only [id_eq]
+  nth_rw 1 [← f.erase_add_single t]
+  rw [mul_add, single_mul_single, Finsupp.add_apply, Finsupp.single_eq_same]
+  convert zero_add _
+  refine Finsupp.not_mem_support_iff.mp (fun h ↦ ?_)
+  refine not_not.mpr ((support_mul (Finsupp.single a r) (f.erase t)) h) ?_
+  simp only [Finsupp.mem_support_single, Finset.mem_biUnion, Ne.def, Finset.mem_singleton,
+    exists_prop, not_exists, not_and, and_imp, forall_eq, Finsupp.support_erase]
+  intros _ x xs
+  refine (add_lt_add_left (WithBot.coe_lt_coe.mp ?_) _).ne'
+  refine WithBot.coe_lt_coe.mpr ((ft _ (Finset.mem_of_mem_erase xs)).lt_of_ne ?_)
+  exact Finset.ne_of_mem_erase xs
```

Comment:
I feel like I reviewed this lemma before in a mathib3 PR, and thought there might be some overlap with a lemma elsewhere. Do you remember which one?

Reply:
Eric, I also know that we already danced this dance.  I do not remember where it was, but I tried to remember what you had suggested for this re-PR, failed to recall what your approach was and re-did it the way I knew... :facepalm: 

It must have been one of the mathlib3 PRs around [mathlib3's #16157](https://github.com/leanprover-community/mathlib/pull/16157), but I failed to find where exactly.

Reply:
Anyway, Junyan reminded me that I actually had introduced `UniqueSums` with the idea of proving a more general version of this instance, so it may not be needed to continue with this PR: I'll try to put something together with UniqueSums tomorrow and, if I am happy with that, will close this PR.